### PR TITLE
[Docs] Add DR covering high-level pager design

### LINKED ISSUE #936: [Design] Paging mechanism design
## What

Design the common paging mechanism for OpenAssetIO methods that work with paging instead of batching

## Requirements

- The API should be as simple as possible to implement in a host and manager, as a view on a fixes set of results. This is not a generic db abstraction.
  - It should be simple to adapt to a broad range of manager implementations as the back-end is unknown
  - It should encourage hosts to consistently implement the API with minimal boilerplate/overhead/foot guns
- The host is required to express in advance the (maximum) number of results required from a query to facilitate optimal implementation:
  - To avoid redundant work when it knows it only needs a maximum of n results.
  - To avoid excessive data transfer in any part of the fulfilment chain.
- The host code path should be as simple as possible to retrieve the requested data to avoid excessive boilerplate and or incorrect handling of data windows.
- We need to be able to communicate consistently when work is done in relation to the invocation of the initial API method, plus any subsequent data window mechanisms.
- The host must be aware that the method it is invoking has scale considerations and can't easily ignore the mechanism by which result size is managed.

- The main API method that generates a cursor must be thread-safe
- The manager must be able to know when a host has finished with any particular cursor to allow resource cleanup.
- The manager must not be expected to fulfill arbitrary data set queries eg: num results, num pages, etc... as these are prohibitive to implement for many back ends.
- The API should be as consistent and coherent with the other OpenAssetIO methods
- Multiple calls to the initial API method must be independent (eg: there is no correlation between two calls to `entityVersions`)

## Scenarios to explore

Timebox: 3 days

- Bulk processing (non UI): I want to iterate all immediate children under an asset and collect their names some other traits (or similar)
- Version popup (UI): I want to display a versions UI with 20 items max, and a "more" menu that pages other versions

- Callbacks .vs. returns (especially considering coherence with the rest of the API + future concerns such as async)
- Single-item vs page array
- Considerations for auditing/debugging, host API isolation, etc.
- Mutability of API objects. Thread safety

## ACs

 - [x] Design for canonical result paging solution that can be used for a method that produces potentially large number of some value-type `T`.
 - [x] Sketches showing it's use explored scenarios (host and manager side)
 - [x] Rationale for chosen/rejected alternatives to facilitate future updates to `CODING_STANDARDS.md` or similar.
 - [x] Specification around the stated API contract in terms of when work is done, re-entrancy, etc...

### PR DESCRIPTION
**Pull request by foundrytom**
Part of #936 

Pretty version [here](https://github.com/OpenAssetIO/OpenAssetIO/blob/c7996576ba27f8f078ac41dc006b6f93dd3449cd/doc/decisions/DR018-Result-pager-design.md).

### CODE REVIEW

File: doc/decisions/DR018-Result-pager-design.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  interactions with the manager are then presented via its native UI.
+
+## Options considered
+
+### Option 1 - "Simple queue"
+
+The pager is a marshal manning a simple queue. There is a large a number
+of results, and they are supplied to the caller a page at a time. Once
+`next` has been called, there is no way to retrieve the data
+for that page again.
```
Comment (elliotcmorris) [inline doc/decisions/DR018-Result-pager-design.md:47]: [Consider]
```suggestion
`next` has been called, there is no way to retrieve the data
for that page again, short of starting again from the very first page.
```

File: doc/decisions/DR018-Result-pager-design.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        Determines if there are results before the currrent page.
+        """
+```
+
+#### Pros
+
+- Facilitates more complex uses of the underlying data beyond gated
+  consumption of a fixed stream.
+- Allows runtimes without delegated UX support to build more advanced
+  workflows.
```
Comment (elliotcmorris) [inline doc/decisions/DR018-Result-pager-design.md:127]: [Consider]
```suggestion
- Allows runtimes without delegated UX support to build more advanced
  workflows.
- Extendable to more complex traversal concepts such as jumps and
  skips should the need for them arise.
```

**Review by elliotcmorris — APPROVED** [review]

File: doc/decisions/DR018-Result-pager-design.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+- Smaller surface are makes it easier to support across a variety of
+  back end implementations.
+- Discourages direct use of the API for complex workflows, encouraging
+  delegated UX, which can result in a more consistent user experience.
+
+#### Cons
+
+- Accessing data from anything other than the first page requires
+  traversal of all preceding pages.
+- Not all runtime environments support delegated UX and/or have critical
```
Comment (feltech) [inline doc/decisions/DR018-Result-pager-design.md:78]: [consider] This confused me a bit at first.  Perhaps worth pointing out in the description of this option that it is assumed UI delegation will pick up the slack for anything more complex.

**Review by feltech — APPROVED** [review]
LGTM

File: doc/decisions/DR018-Result-pager-design.md (hunk 3 above)
Comment (foundrytom) [reply doc/decisions/DR018-Result-pager-design.md:82]: Added [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/985/commits/9fdea11e410d40fd0e32bd21e29890b679c37009).

File: doc/decisions/DR018-Result-pager-design.md (hunk 1 above)
Comment (foundrytom) [reply doc/decisions/DR018-Result-pager-design.md]: Added [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/985/commits/a84feb39ca9dbe926f4c15c1cfeaf48d6eb3477e).

File: doc/decisions/DR018-Result-pager-design.md (hunk 2 above)
Comment (elliotcmorris) [reply doc/decisions/DR018-Result-pager-design.md:131]: added [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/985/commits/f91090f12ec180e79114258d32acbc31622a98ef)

Comment (foundrytom) [reply doc/decisions/DR018-Result-pager-design.md:131]: 🤔 Sorry - was sure I added a comment to that effect!

Comment (foundrytom) [reply doc/decisions/DR018-Result-pager-design.md:131]: Oh, its as I accepted your suggestion so it auto closes the thread.

**Review by feltech — APPROVED** [review]

**Pull request merged**